### PR TITLE
feat: allows spawned peds to hold a weapon

### PR DESCRIPTION
**Pull request by Spudgun**
as title says

### CODE REVIEW

**Comment by BerkieBb** [comment]
Please use `joaat` instead of `GetHashKey`

**Comment by Spudgun** [comment]
The if statement that has `GetHashKey` only runs if the user doesn't `jooat` the weapon name with backticks in the `Config.Ped` table under `init.lua` file.

`GetHashKey` is used in the same way about 7 other times in the `client.lua` file. Essentially, just a redundancy check as this is a public script with multiple users of differing coding experience. Happy to edit whatever needs to be edited but figured I'd keep the same checks you have put in place for this as it is a good idea 👍🏽

**Comment by BerkieBb** [comment]
As I said, please use `joaat` instead of `GetHashKey`, it's not used anymore in qb-target at all, you're running an older version, `joaat` is the faster alternative of `GetHashKey` with it being faster because it runs at runtime instead of invoking a native

**Comment by Spudgun** [comment]
> As I said, please use `joaat` instead of `GetHashKey`, it's not used anymore in qb-target at all, you're running an older version, `joaat` is the faster alternative of `GetHashKey` with it being faster because it runs at runtime instead of invoking a native

Apologies.. we were working with an outdated version. I'll pull down the changes tomorrow and update the PR to include `jooat` 😊

**Comment by BerkieBb** [comment]
I saw the commit, this may sound a little weird that I check that or care that much, but you left some trailing whitespace with the latest commit

**Comment by Spudgun** [comment]
> I saw the commit, this may sound a little weird that I check that or care that much, but you left some trailing whitespace with the latest commit

That's completely fair. Should be removed now

**Comment by BerkieBb** [comment]
I hope I'm not asking too much, but could you document this aswell in the TEMPLATES.md, you can just add this option to the existing documentation to it

**Comment by Spudgun** [comment]
Sure thing, will go and add the documentation now

**Comment by Spudgun** [comment]
Apologies about so many doc changes... just realized we could add target options for ped spawn 🤦‍♂️

**Comment by BerkieBb** [comment]
No problem, but could you also add the options to the TEMPLATES.md, I asked for that originally, but I don't mind you adding more docs, it looks great

**Comment by Spudgun** [comment]
Should be good to go now.  However, if anything else needs changing just let me know.

Also, quick question...  have you thought about assigning ped relationship groups and the interaction between groups or players to peds? The logic behind it is a bit more complex but could be a nice addition

**Comment by BerkieBb** [comment]
I haven't thought about that, but feel free to add it to this PR, also, currently the weapon logic is only implemented for the SpawnPeds function which takes peds from the config, but there is also a SpawnPed function which is an export, could you add the logic to that aswell? (Same for the TEMPLATES.md, the export is located under the config option there, also the format should have that table in there)

**Comment by Spudgun** [comment]
I'll have a play around with the relationship stuff over the next few days. Just have to give it a bit of a think-over on how best to implement the relationships. 

I've just added functionality for the SpawnPed export and also added the documentation into `Templates.md`

**Comment by BerkieBb** [comment]
One thing left for the documentation, the format, it's located above the config option

**Comment by Spudgun** [comment]
I've added that into the documentation for the `TEMPLATES.md`.

In regards to the relationship group... I believe the best way to do this would be to set the `SetPedRelationshipGroupHash` for the individual ped created. You could then add their relationship to other groups but I feel like this is a little excessive as it's not localized to just that ped, but rather the group that they get assigned. Meaning multiple instances of assignment could potentially happen upon ped creation. Example below..

**Ped #1**
```lua
pedrelations = {
    group = 'mrpdguards',
    relationshiptogroups = {
        {
            group = 'AMBIENT_GANG_BALLAS'
            type = 1,
        },
        {
            group = 'AMBIENT_GANG_FAMILY '
            relationship = 2,  
        },
    }		
},
```

**Ped #2**
```lua
pedrelations = {
    group = 'mrpdguards',
    relationshiptogroups = {
        {
            group = 'AMBIENT_GANG_BALLAS'
            type = 1,
        },
        {
            group = 'AMBIENT_GANG_FAMILY '
            type = 2,  
        },
    }		
},
```

```lua
if v.pedrelations and type(v.pedrelations.group) == 'string' then
    local grouphash = joaat(v.pedrelations.group) 

    if not DoesRelationshipGroupExist(grouphash) then
        AddRelationshipGroup(v.pedrelations.group)
    end

    SetPedRelationshipGroupHash(spawnedped, grouphash)

    --[[
    ####    This code below is semi redundant if multiple peds are created with
    ####    the same relationship group and have the same relationships between
    ####    different groups.
    ]]--

    -- for _, v2 in pairs(v.pedrelations.relationshiptogroups) do
    --     if type(v2.group) == 'string' then v2.group = jooat(v2.group) end

    --     if DoesRelationshipGroupExist(v2.group) then
    --         SetRelationshipBetweenGroups(v.type, grouphash, v2.group)
    --     end					
    -- end
else
    if Config.Debug then print(v.pedrelations.group .. ' is not a string') end
end
```

I believe a better way to do this would be either to create an `action` value in the config that has a function attached that runs on ped creation so the player can utilize that to call events etc or to `return` the ped that is created via the **export** `SpawnPed`.

The first option covers all use cases I believe. Users entering peds into the `Config.Peds` in `init.lua` and also the **export** option if they choose to.

 Let me know your thoughts 😊

**Comment by BerkieBb** [comment]
https://github.com/citizenfx/lua/tree/luaglm-dev/cfx
Reference this readme btw, for lua5.4 and lua oal adjustments from cfxlua

**Comment by Spudgun** [comment]
Thanks for this. I'll save that documentation for reference.

I have now added in ped relations in the latest commit. I have also added in an action function or when a ped is created in case the user would like to trigger any events or run any functions upon ped spawn.

File: client.lua [hunk 1]
```diff
@@ -808,6 +841,35 @@ local function SpawnPed(data)
...
+					SetPedRelationshipGroupHash(spawnedped, pedgrouphash)
+					if v.pedrelations.toplayer then
+						SetRelationshipBetweenGroups(v.pedrelations.toplayer, pedgrouphash, joaat('PLAYER'))
+					end
+
+					if v.pedrelations.toowngroup then
+						SetRelationshipBetweenGroups(v.pedrelations.toowngroup, pedgrouphash, pedgrouphash)
+					end
+				else
+					if Config.Debug then print(v.pedrelations.groupname .. ' is not a string') end
```
Comment (BerkieBb) [inline client.lua:860]: Please make this throw an error by using `error()`

File: client.lua [hunk 2]
```diff
@@ -871,6 +937,35 @@ local function SpawnPed(data)
...
+				SetPedRelationshipGroupHash(spawnedped, pedgrouphash)
+				if data.pedrelations.toplayer then
+					SetRelationshipBetweenGroups(data.pedrelations.toplayer, pedgrouphash, joaat('PLAYER'))
+				end
+
+				if data.pedrelations.toowngroup then
+					SetRelationshipBetweenGroups(data.pedrelations.toowngroup, pedgrouphash, pedgrouphash)
+				end
+			else
+				if Config.Debug then print(data.pedrelations.groupname .. ' is not a string') end
```
Comment (BerkieBb) [inline client.lua:956]: Make an error here aswell using `error()`

File: client.lua [hunk 3]
```diff
@@ -732,6 +732,35 @@ function SpawnPeds()
...
+				SetPedRelationshipGroupHash(spawnedped, pedgrouphash)
+				if v.pedrelations.toplayer then
+					SetRelationshipBetweenGroups(v.pedrelations.toplayer, pedgrouphash, joaat('PLAYER'))
+				end
+
+				if v.pedrelations.toowngroup then
+					SetRelationshipBetweenGroups(v.pedrelations.toowngroup, pedgrouphash, pedgrouphash)
+				end
+			else
+				if Config.Debug then print(v.pedrelations.groupname .. ' is not a string') end
```
Comment (BerkieBb) [inline client.lua:751]: Make it error out here using `error()`

**Review by BerkieBb — CHANGES_REQUESTED** [review]

File: client.lua (hunk 1 above)
Comment (Spudgun) [reply client.lua]: All done

File: client.lua (hunk 2 above)
Comment (Spudgun) [reply client.lua]: All done

File: client.lua (hunk 3 above)
Comment (Spudgun) [reply client.lua]: All done

**Comment by BerkieBb** [comment]
Could you make a new pull request to clear up the history? Then it should be good to go

**Comment by Spudgun** [comment]
Sure thing, will do that now!

**Pull request closed without merging**